Chia-Network/chia-blockchain
# exempt_peer_networks: allow exceeding target_peer_count.

**Background**: I would like to be able to allow specific IPs/IP ranges to always make connections to my full node, irrespective of whether I have already reached `target_peer_count` connected peers.

**Proposal**: new configuration option `exempt_peer_networks` including a list of networks or individual IP addresses allowed to connect even when `target_peer_count` has been reached (ie networks where peers should be exempt from connection limits).

Config example:

```
  # IPv4/IPv6 network addresses/CIDR blocks allowed to connect even when target_peer_count has been hit.
  # exempt_peer_networks = ["192.168.0.3", "192.168.1.1/24", "fe80::/10", "2606:4700:4700::64/128"]
  exempt_peer_networks:
    - "10.0.2.15"
    - "192.168.1.0/24"
    - "100.64.0.0/10"
    - "fe80::/10"
```


`ipaddress.ip_network` module supports input of both IPv4/IPv6 CIDR blocks and individual IPv4/IPv6 addresses (assumes `/32` or `/128`).  Invalid configs will throw an exception at startup time.   

**Impact** This makes it possible to run a full node (farmer, $5 digital ocean droplet) remotely and always allow connections from certain network addresses (trusted servers on the same network, laptop GUI at the office, etc) and ensure trusted peers can leverage that full node to sync quickly.

**How to test**

1. Add an `exempt_peer_networks` block to your full_node config.
2. Restart your full node, and wait until it has a full complement on connections:`chia show -c  |grep -c FULL_NODE` shows 50.
3. Open a GUI full node elsewhere, click "Connect to Other Peers" and enter the IP your full node.
4. Note that `chia show -c  |grep -c FULL_NODE`  now shows 51 and that the connection worked.

*Risks*:  Denial of service/Resource exhaustion/network de-sync if you set an overly broad `exempt_peer_networks` (e.g. "0.0.0.0/0") or attacker can connect from many addresses in an allowed range and use all `target_peer_count` (default=60) incoming connections. 

This is my first PR, please let me know if you have any feedback.  I couldn't easily come up with a way to write tests but am open to suggestions.
